### PR TITLE
Refactor plugin architecture

### PR DESCRIPTION
**Pull request by schaefi**
Set kiwi.tasks to be the plugin entry point and register
existing task plugins in setup.py. Change the code in
cli.py to auto register plugins using the iter_entry_points
method from the pkg_resources class. This allows for easier
writing of external kiwi plugins.

### CODE REVIEW

**Comment by schaefi** [comment]
Inspired by https://packaging.python.org/guides/creating-and-discovering-plugins/ I created this PR. It changes the existing task plugin code to allow for external plugins if installed to the kiwi.tasks package.

This is just a start. We are definitely missing a documentation example. Some code parts were cleaned up by that change.

I have not done an example plugin with an installed python-kiwi outside of a virtual environment. Maybe I missed some parts, but this is open for a review now

Thanks

File: setup.py [hunk 1]
```diff
@@ -199,6 +201,16 @@ def run(self):
         'clean': clean
     },
     'entry_points': {
+        'kiwi.tasks': [
```
Comment (davidcassany) [inline setup.py]: I guess having them all listed here is the counter part of the refactor of cli.py based on `pkg_resources`. I did not expect that, but I understand this is better than having to parse the OS file within `kiwi/tasks` folder.

Comment (schaefi) [reply setup.py]: yeah let's keep this part explicit. I actually also like it better, it's not a real hurdle

**Comment by schaefi** [comment]
> in any case I can try to port my plugin to some repo and make a trial including this kiwi version as a dependency. 

I think this would be good. I'd like to add some sort of documentation before we merge it.
I'll keep this open until you have time to send feedback if adding external plugins is now more straight forward. Thanks much

**Comment by schaefi** [comment]
@davidcassany I have created a chapter to describe the plugin architecture including an example. Could you read and verify that it really works that way ?

Thanks much

**Comment by schaefi** [comment]
@tomschr if you have a minute to look at the new doc chapter from a documentation point of view it would be great :) Thanks

File: doc/source/development/kiwi_plugin_architecture.rst [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+   The following example assumes an existing python project
+   exists and was setup according to the python project rules
+   and standards.
+
+1. Create the task plugin directory
+
+   .. code:: bash
+
+       $ mkdir -p kiwi/tasks
```
[the remark this answers is not in the document]
Comment (schaefi) [reply doc/source/development/kiwi_plugin_architecture.rst]: LOL, last famous words

[the remark this answers is not in the document]
Comment (schaefi) [reply doc/source/development/kiwi_plugin_architecture.rst]: I will change it

File: doc/source/development/kiwi_plugin_architecture.rst [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+          'kiwi.tasks': [
+              'relax_justdoit=kiwi.tasks.relax_justdoit'
+          ]
+      }
+
+3. Create the plugin code
+
+   .. code:: bash
+ 
+       $ touch kiwi/tasks/relax_justdoit.py
```
[the remark this answers is not in the document]
Comment (schaefi) [reply doc/source/development/kiwi_plugin_architecture.rst]: Thanks it really took me a while to find something with a fun factor :-)

File: doc/source/development/kiwi_plugin_architecture.rst [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+               print(
+                   'https://genius.com/Frankie-goes-to-hollywood-relax-lyrics'
+               )
+
+4. Test the plugin
+
+   .. code:: bash
+
+       $ python ./setup.py develop
```
[the remark this answers is not in the document]
Comment (schaefi) [reply doc/source/development/kiwi_plugin_architecture.rst]: yep let's go with the direct invocation model

**Comment by davidcassany** [comment]
I can confirm this works fine. I created a python package for kiwi with `./setup.py sdist` from my local development environment for this current branch. Then in an independent virtual environment I manually installed this KIWI package with `pip install ./kiwi-9.20.1.tar.gz`. After that I could easily create a new task for kiwi by just using the setuptools as documented here. The `setup.py` file I used to the plugin looks like (omitting most of the metadata):

```python
#!/usr/bin/env python
# -*- encoding: utf-8 -*-

from setuptools import setup

config = {
    'name': 'myKIWIplugin',
    'description': 'wonderful KIWI plugin',
    'version': '0.1',
    'license' : 'GPLv3+',
    'install_requires': [
        'kiwi >= 9.20.1'
    ],
    'packages': ['my_plugin'],
    'entry_points': {
        'kiwi.tasks': [
            'system_newbuild=my_plugin.system_newbuild',
        ]
    }
}

setup(**config)
```

**Pull request merged**